core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -470,12 +470,25 @@ class SparkContext(config: SparkConf) extends Logging {
       files.foreach(addFile)
     }
 
-    _executorMemory = _conf.getOption("spark.executor.memory")
-      .orElse(Option(System.getenv("SPARK_EXECUTOR_MEMORY")))
-      .orElse(Option(System.getenv("SPARK_MEM"))
-      .map(warnSparkMem))
-      .map(Utils.memoryStringToMb)
-      .getOrElse(1024)
+    _executorMemory = {
+      val defaultMemory = 1024
+      val configuredMemory = _conf.getOption("spark.executor.memory")
+        .orElse(Option(System.getenv("SPARK_EXECUTOR_MEMORY")))
+        .orElse(Option(System.getenv("SPARK_MEM")).map(warnSparkMem))
+        .map(Utils.memoryStringToMb)
+      // In local-cluster mode, always use the slave memory specified in the master string
+      // In other modes, use the configured memory if it exists
+      master match {
+        case SparkMasterRegex.LOCAL_CLUSTER_REGEX(_, _, em) =>
+          if (configuredMemory.isDefined) {
```

Comment:
Could you at least change this so that `spark.executor.memory` takes precedence if it's set? Then both use cases are possible. (Maybe someone is crazy enough to be trying dynamic allocation in local-cluster mode, or something else...)